cpp/include/cudf/io/datasource.hpp

```diff
@@ -36,6 +37,212 @@ namespace io {
  * @file
  */
 
+/**
+ * @brief Kind of data source to create when calling
+ * `cudf::io::datasource::create()`.
+ *
+ * @see cudf::io::datasource::create()
+ * @see cudf::io::datasource_params
+ *
+ * N.B. GDS = GPUDirect Storage
+ */
+enum class datasource_kind {
+  /**
+   * @brief Kvikio-based data source (default).
+   *
+   * This data source is the default for cuDF, and should be the most performant
+   * option for most use cases.  It supports GDS where possible, falling back to
+   * multi-threaded host-based reads when GDS is not available.
+   *
+   * It supports asynchronous reads, and will use the provided CUDA stream for
+   * all I/O operations when possible.
+   */
+  KVIKIO = 0,
+  DEFAULT = KVIKIO,
```

Comment:
We're considering having a dynamic default - use different implementation based on file or system parameters. I don't want to lock the default down in this API.
I'm thinking that we could default to something like `AUTO` here, or pass `datasource_kind` as optional.

Reply:
Yeah `AUTO` or some not-necessarily-`KVIKIO` default would be fine.  As it currently stands... there's no way of getting the existing behavior, i.e. the `file_source` backend, which... seems like something that should be reviewed.

So that also means there's no way of getting as simple `cuFile`-based implementation... a `CUFILE` datasource kind would be handy.

I also played around with an `io_uring` implementation... doesn't play nicely with GDS obviously, but should provide the fastest host-based reader if done correctly.  Not including that in this patch, too much scope creep.

Overall it sounds like the parameterizing of `datasource_kind` and `datasource_params` as an API change is reasonable, though?  The other option would be to expose all the `*_source` classes and let callers create them directly.